tactic/back.lean

```diff
@@ -0,0 +1,121 @@
+-- Copyright (c) 2017 Scott Morrison. All rights reserved.
+-- Released under Apache 2.0 license as described in the file LICENSE.
+-- Authors: Scott Morrison, Keeley Hoek
+
+import tactic.basic tactic.ext
+
+namespace tactic
+
+@[user_attribute]
+meta def back_attribute : user_attribute unit (option unit) := {
+  name := `back,
+  descr := "A lemma that should be applied to a goal whenever possible; use `back` to automatically `apply` all lemmas tagged `[back]`. Lemmas tagged with `[back!]` may only be applied during the `solve_by_elim` finishing step of `back`.",
+  parser := optional $ lean.parser.tk "!"
+}
+
+/-- Stub for implementing faster lemma filtering using Pi binders in the goal -/
+private meta def is_lemma_applicable (lem : expr) : tactic bool := return true
+
+/-- Takes two sets of lemmas, 'finishing' lemmas and 'progress' lemmas.
+
+    Finishing lemmas should be applied only as part of a sequence of applications that close the goals.
+    Progress lemmas should be applied whenever possible, regardless of new hypotheses.
+
+    Finishing lemmas are applied in preference to progress lemmas.
+
+    `back` succeeds if at least one progress lemma was applied, or all goals are discharged.
+    -/
+
+-- TODO: Simon Hudon and Scott Morrison have discussed speeding this up, by indexing.
+meta def back (solve_by_elim_desc : string) (finishing_lemmas : list expr) :
+  list expr → bool → tactic (list string)
+| progress_lemmas progress :=
+  -- Note: the order here matters: we should prioritise using assumptions,
+  -- over further applications of progress lemmas.
+  (do solve_by_elim_aux done (return (finishing_lemmas ++ progress_lemmas)) 5,
+      return [solve_by_elim_desc]) <|>
+  (do let candidate_lemmas := progress_lemmas.mfilter is_lemma_applicable,
+      q ← (progress_lemmas.any_of $ λ e, apply_thorough e >>= (λ p, pure (e, p.2))),
+      r' ← back (progress_lemmas.erase q.1) tt,
+      return (q.2 ++ r')) <|>
+  guard progress >> pure []
+
+private meta def lookup_tagged_lemma (n : name) : tactic (bool × expr) :=
+do is_elim ← back_attribute.get_param n,
+   e ← mk_const n,
+   return (is_elim.is_some, e)
+
+private meta def collect_tagged_lemmas : list name → tactic (list expr × list expr)
+| [] := return ([], [])
+| (n :: rest) := do (normals, elims) ← collect_tagged_lemmas rest,
+                    (is_elim, e) ← lookup_tagged_lemma n,
+                    return $ if is_elim then (normals, e :: elims) else (e :: normals, elims)
+
+private meta def mk_default_finishing_lemmas : tactic (list expr) :=
+do congr_fun ← mk_const `congr_fun,
+   congr_arg ← mk_const `congr_arg,
+   return [congr_fun, congr_arg]
+
+/--
+Returns a list of "finishing lemmas" (which should only be applied as part of a
+chain of applications which discharges the goal), and a list of "progress lemmas",
+the successful application of any one of which counting as a success.
+
+This is a close relative of `mk_assumption_set` in tactic/basic.lean.
```

Comment:
Does this exist?

Reply:
Ah, sorry, it's in `tactic.interactive.lean`.

Reply:
I've fixed this.